cases/validate-value.ts

```diff
@@ -0,0 +1,47 @@
+import { isOfType, Value } from 'validate-value';
+import { Data } from '../data';
+import { Case } from './abstract';
+
+const dataType = new Value({
+  type: 'object',
+  required: [
+    'boolean',
+    'deeplyNested',
+    'longString',
+    'maxNumber',
+    'negNumber',
+    'number',
+    'string',
+  ],
+  properties: {
+    number: { type: 'number' },
+    negNumber: { type: 'number' },
+    maxNumber: { type: 'number' },
+    string: { type: 'string' },
+    longString: { type: 'string' },
+    boolean: { type: 'boolean' },
+    deeplyNested: {
+      type: 'object',
+      required: ['foo', 'bool', 'num'],
+      properties: {
+        foo: { type: 'string' },
+        num: { type: 'number' },
+        bool: { type: 'boolean' },
+      },
+    },
+  },
+});
+
+export class ValidateValueCase extends Case implements Case {
+  name = 'validate-value';
+
+  validate() {
+    const { data } = this;
+
+    if (isOfType<Data>(data, dataType.schema)) {
```

Comment:
We also don't want to type hint with `Data`, because that defeats the purpose of DRY. The idea is to define the validation once, and from that TS can infer the type.

Reply:
TypeScript cannot infer the type `isOfType()` should narrow `data` to for the same reason it must be narrowed: `data` is of type `any`. Unfortunately, validate-value provides no type relation between `dataType.schema` and any would-be valid objects conforming to the schema.

Reply:
> TypeScript cannot infer the type

Really? But the docs say that it is a type guard... The only purpose of the type guard is to infer type. So are the docs wrong then?

Reply:
Hm, yeah, ok, I verified it with their example and type guarding indeed does not work. In which case then this library does not fit the parameters of this repository.

Reply:
See criteria: https://github.com/moltar/typescript-runtime-type-benchmarks#criteria

The idea is that we don't want to define the validation AND the type in two places. Types should be inferred from the schema. See all of the other test cases. In no other test case we use the `Data` type to hint what the type is. The validation or type guarding function infers the type automatically.

Reply:
> Really? But the docs say that it is a type guard... The only purpose of the type guard is to infer type. So are the docs wrong then?

It is a generic type guard where you must write the type you wish `dataType.schema` to represent. The library does not evaluate what the expected type is on its own because it does not compute what objects conforming to `dataType.schema` would look like. As a result, the only thing TypeScript understands when `data` passes through the type guard is that the type of `data` conforms to the schema `dataType.schema`.

Reply:
> It is a generic type guard

Fair.

---

But in that case it does not fit the purpose of this repository. So we should close this PR, or fix the library to do type guarding without generics hints.